<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Public · SNaQ.jl</title><meta name="title" content="Public · SNaQ.jl"/><meta property="og:title" content="Public · SNaQ.jl"/><meta property="twitter:title" content="Public · SNaQ.jl"/><meta name="description" content="Documentation for SNaQ.jl."/><meta property="og:description" content="Documentation for SNaQ.jl."/><meta property="twitter:description" content="Documentation for SNaQ.jl."/><meta property="og:url" content="https://JuliaPhylo.github.io/SNaQ.jl/lib/public/"/><meta property="twitter:url" content="https://JuliaPhylo.github.io/SNaQ.jl/lib/public/"/><link rel="canonical" href="https://JuliaPhylo.github.io/SNaQ.jl/lib/public/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="SNaQ.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SNaQ.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/installation/">Installation</a></li><li><a class="tocitem" href="../../man/snaq_est/">Network estimation</a></li><li><a class="tocitem" href="../../man/fixednetworkoptim/">Candidate networks</a></li><li><a class="tocitem" href="../../man/expectedCFs/">Extract expected CFs</a></li><li><a class="tocitem" href="../../man/bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../../man/parallelcomputation/">Improving runtimes</a></li><li><a class="tocitem" href="../../man/multiplealleles/">Multiple alleles</a></li><li><a class="tocitem" href="../../man/error_reporting/">Error reporting</a></li></ul></li><li><span class="tocitem">Library</span><ul><li class="is-active"><a class="tocitem" href>Public</a><ul class="internal"><li><a class="tocitem" href="#Functions-and-Types"><span>Functions &amp; Types</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li></ul></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Public</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Public</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/SNaQ.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/dev/docs/temp_src_for_build/lib/public.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><div class="admonition is-info" id="Important-Note:-6724263f5023a524"><header class="admonition-header">Important Note:<a class="admonition-anchor" href="#Important-Note:-6724263f5023a524" title="Permalink"></a></header><div class="admonition-body"><p>This documentation pertains to SNaQ v1.1 and may differ from the specific implementation originally described in <a href="https://doi.org/10.1371/journal.pgen.1005896">Solís-Lemus &amp; Ané (2016)</a>. See documentation SNaQ v1.0  for the original implementation.</p></div></div><h1 id="Public-Documentation"><a class="docs-heading-anchor" href="#Public-Documentation">Public Documentation</a><a id="Public-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#Public-Documentation" title="Permalink"></a></h1><p>Documentation for <code>SNaQ</code>&#39;s public (exported) interface.</p><h2 id="Functions-and-Types"><a class="docs-heading-anchor" href="#Functions-and-Types">Functions &amp; Types</a><a id="Functions-and-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-and-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.bootsnaq-Tuple{HybridNetwork, Union{DataFrames.DataFrame, Vector{Vector{HybridNetwork}}}}" href="#SNaQ.bootsnaq-Tuple{HybridNetwork, Union{DataFrames.DataFrame, Vector{Vector{HybridNetwork}}}}"><code>SNaQ.bootsnaq</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">bootsnaq(T::HybridNetwork, df::DataFrame)
bootsnaq(T::HybridNetwork, vector of tree lists)</code></pre><p>Bootstrap analysis for SNaQ. Bootstrap data can be quartet concordance factors (CF), drawn from sampling uniformly in their credibility intervals, as given in the data frame <code>df</code>. Alternatively, bootstrap data can be gene trees sampled from a vector of tree lists: one list of bootstrap trees per locus (see <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.readmultinewick_files-Tuple{AbstractString}"><code>readmultinewick_files</code></a> in <code>PhyloNetworks</code> to generate this, from a file containing a list of bootstrap files: one per locus).</p><p>From each bootstrap replicate, a network is estimated with <a href="#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>snaq!</code></a>, with a search starting from topology <code>T</code>. Optional arguments include the following, with default values in parentheses:</p><ul><li><code>hmax</code> (1): max number of reticulations in the estimated networks</li><li><code>nrep</code> (10): number of bootstrap replicates.</li><li><code>runs</code> (10): number of independent optimization runs for each replicate</li><li><code>filename</code> (&quot;bootsnaq&quot;): root name for output files. No output files if &quot;&quot;.</li><li><code>seed</code> (0 to get a random seed from the clock): seed for random number generator</li><li><code>otherNet</code> (empty): another starting topology so that each replicate will start prcnet% runs on otherNet and (1-prcnet)% runs on <code>T</code></li><li><code>prcnet</code> (0): percentage of runs starting on <code>otherNet</code>; error if different than 0.0, and otherNet not specified.</li><li><code>ftolRel</code>, <code>ftolAbs</code>, <code>xtolRel</code>, <code>xtolAbs</code>, <code>liktolAbs</code>, <code>Nfail</code>, <code>probST</code>, <code>verbose</code>, <code>outgroup</code>: see <code>snaq!</code>, same defaults.</li></ul><p>If <code>T</code> is a tree, its branch lengths are first optimized roughly with <a href="../internals/#SNaQ.updateBL!-Tuple{HybridNetwork, DataCF}"><code>updateBL!</code></a> (by using the average CF of all quartets defining each branch and calculating the coalescent units corresponding to this quartet CF). If <code>T</code> has one or more reticulations, its branch lengths are taken as is to start the search. The branch lengths of <code>otherNet</code> are always taken as is to start the search.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/bootstrap.jl#L219-L251">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.fittedquartetCF" href="#SNaQ.fittedquartetCF"><code>SNaQ.fittedquartetCF</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">fittedquartetCF(d::DataCF, format::Symbol)</code></pre><p>Data frame with the observed and expected quartet concordance factors after estimation of a network with <code>snaq!</code>, or fitting of quartet CF data on a fixed network. The format can be <code>:wide</code> (default) or <code>:long</code>.</p><ul><li>if <code>wide</code>, the output has one row per 4-taxon set, and each row has 10 columns: 4 columns for the taxon names, 3 columns for the observed CFs and 3 columns for the expected CF.</li><li>if <code>long</code>, the output has one row per quartet, i.e. 3 rows per 4-taxon sets, and 7 columns: 4 columns for the taxon names, one column to give the quartet resolution, one column for the observed CF and the last column for the expected CF.</li></ul><p>see also: <a href="#SNaQ.topologyQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>topologyQpseudolik!</code></a> and <a href="#SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>topologymaxQpseudolik!</code></a> to update the fitted quartet CF expected under a specific network, inside the DataCF object <code>d</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/descriptive.jl#L6-L23">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.mapallelesCFtable-Tuple{AbstractString, AbstractString}" href="#SNaQ.mapallelesCFtable-Tuple{AbstractString, AbstractString}"><code>SNaQ.mapallelesCFtable</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mapallelesCFtable(mapping file, CF file; filename, columns, delim)</code></pre><p>Create a new DataFrame containing the same concordance factors as in the input CF file, but with modified taxon names. Each allele name in the input CF table is replaced by the species name that the allele maps onto, based on the mapping file. The mapping file should have column names: allele and species.</p><p>Optional arguments:</p><ul><li><code>filename</code> to write/save resulting CF table. If not specified, then the output data frame is not saved to a file.</li><li><code>columns</code> column numbers for the taxon names. 1-4 by default.</li><li>any keyword arguments that <code>CSV.File</code> would accept. For example, <code>delim=&#39;,&#39;</code> by default: columns are delimited by commas. Unless specified otherwise by the user, <code>pool=false</code> (to read taxon names as Strings, not levels of a categorical factor, for combining the 4 columns with taxon names more easily). The same CSV arguments are used to read both input file (mapping file and quartet file)</li></ul><p>See also <a href="../internals/#SNaQ.mapallelesCFtable!-Tuple{DataFrames.DataFrame, DataFrames.DataFrame, AbstractVector{Int64}, Bool, AbstractString}"><code>mapallelesCFtable!</code></a> to input DataFrames instead of file names.</p><p>If a <code>filename</code> is specified, such as &quot;quartetCF_speciesNames.csv&quot; in the example below, this file is best read later with the option <code>pool=false</code>. example:</p><pre><code class="language-julia hljs">mapallelesCFtable(&quot;allele-species-map.csv&quot;, &quot;allele-quartet-CF.csv&quot;;
                  filename = &quot;quartetCF_speciesNames.csv&quot;)
df_sp = CSV.read(&quot;quartetCF_speciesNames.csv&quot;, DataFrame); # DataFrame object
dataCF_specieslevel = readtableCF!(df_sp, mergerows=true); # DataCF object</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/multipleAlleles.jl#L8-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readmultinewicklevel1-Tuple{AbstractString}" href="#SNaQ.readmultinewicklevel1-Tuple{AbstractString}"><code>SNaQ.readmultinewicklevel1</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readmultinewicklevel1(file)</code></pre><p>Read a text file with a list of trees/networks in extended newick format (one tree per line) and transform them like <a href="#SNaQ.readnewicklevel1-Tuple{AbstractString}"><code>readnewicklevel1</code></a>. Namely, in each tree/network</p><ul><li>the root is suppressed (becomes of degree 3 if it was of degree 2)</li><li>any polytomy is resolved arbitrarily</li><li>any missing branch length is set to 1</li><li>any branch length above 10 is set to 10 (this assumes branch lengths in coalescent units)</li><li>any missing γ&#39;s are set to (0.1, 0.9)</li></ul><p>and more (see <a href="#SNaQ.readnewicklevel1-Tuple{AbstractString}"><code>readnewicklevel1</code></a>).</p><p>See <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.readmultinewick"><code>PhyloNetworks.readmultinewick</code></a> to read multiple trees or networks with no modification.</p><p>Output: array of HybridNetwork objects.</p><p>Each line starting with &quot;(&quot; will be considered as describing one topology. The file can have extra lines that are ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readquartetdata.jl#L219-L239">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readnewicklevel1-Tuple{AbstractString}" href="#SNaQ.readnewicklevel1-Tuple{AbstractString}"><code>SNaQ.readnewicklevel1</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readnewicklevel1(filename)
readnewicklevel1(parenthetical format)</code></pre><p>Similarly to <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.readnewick-Tuple%7BAbstractString%7D"><code>PhyloNetworks.readnewick</code></a>: read a tree or network in parenthetical format, but this function enforces the necessary conditions for any starting topology in SNaQ: non-intersecting cycles, no polytomies, unrooted. It sets any missing branch length to 1.0, and reduces any branch length above 10 to 10 (as it assumes branch lengths are in coalescent units).</p><p>If the network has a bad diamond II (in which edge lengths are γ&#39;s are not identifiable) and if the edge below this diamond has a length <code>t</code> different from 0, then this length is set back to 0 and the major parent hybrid edge is lengthened by <code>t</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readwrite.jl#L176-L189">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readsnaqnetwork-Tuple{AbstractString}" href="#SNaQ.readsnaqnetwork-Tuple{AbstractString}"><code>SNaQ.readsnaqnetwork</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readsnaqnetwork(output file)</code></pre><p>Read the estimated network from a <code>.out</code> file generated by <a href="#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>snaq!</code></a>. The network score is read also, and stored in the network&#39;s field <code>.fscore</code>.</p><p>Warning: the score is not the log likelihood and is instead proportional to the network&#39;s pseudo-deviance: the lower, the better. Do NOT use this score to calculate an AIC or BIC (etc.) value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readwrite.jl#L409-L418">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readtableCF!-Tuple{DataCF, DataFrames.DataFrame, Vector{Int64}}" href="#SNaQ.readtableCF!-Tuple{DataCF, DataFrames.DataFrame, Vector{Int64}}"><code>SNaQ.readtableCF!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readtableCF!(data frame, columns; mergerows=false)</code></pre><p>Read in quartet CFs from data frame, assuming information is in columns numbered <code>columns</code>, of length <strong>7 or 8</strong>: 4 taxon labels then 3 CFs then ngenes possibly.</p><p>If some species appears more than once in the same 4-taxon set (e.g. t1,t1,t2,t3), then the data frame is modified to remove rows (4-taxon sets) that are uninformative about between-species relationships. This situation may occur if multiple individuals are sampled from the same species. A 4-taxon set is uninformative (and its row is removed) if one taxon is repeated 3 or 4 times (like t1,t1,t1,t1 or t1,t2,t2,t2). The list of species appearing twice in some 4-taxon sets is stored in the output DataCF object. For these species, the length of their external edge is identifiable (in coalescent units). If multiple rows correspond to the same 4-taxon set, these rows are merged and their CF values (and number of genes) are averaged. If none of the species is repeated within any 4-taxon set, then this averaging is attempted only if <code>mergerows</code> is true.</p><pre><code class="language-julia hljs">readtableCF!(DataCF, data frame, columns)</code></pre><p>Modify the <code>.quartet.obsCF</code> values in the <code>DataCF</code> object with those read from the data frame in columns numbered <code>columns</code>. <code>columns</code> should have <strong>3</strong> columns numbers for the 3 CFs in this order: <code>12_34</code>, <code>13_24</code> and <code>14_23</code>.</p><p>Assumptions:</p><ul><li>same 4-taxon sets in <code>DataCF</code> and in the data frame, and in the same order, but this assumption is <em>not checked</em> (for speed, e.g. during bootstrapping).</li><li>one single row per 4-taxon set (multiple individuals representatives of the same 4-taxon set should have been already merged); basically: the DataCF should have been created from the data frame by <code>readtableCF!(df, colums)</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readquartetdata.jl#L176-L207">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readtableCF-Tuple{AbstractString}" href="#SNaQ.readtableCF-Tuple{AbstractString}"><code>SNaQ.readtableCF</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readtableCF(file)
readtableCF(data frame)
readtableCF!(data frame)</code></pre><p>Read a file or DataFrame object containing a table of concordance factors (CF), with one row per 4-taxon set. The first 4 columns are assumed to give the labels of the 4 taxa in each set (tx1, tx2, tx3, tx4). Columns containing the CFs are assumed to be named <code>CF12_34</code>, <code>CF13_24</code> and <code>CF14_23</code>; or <code>CF12.34</code>, <code>CF13.24</code> and <code>CF14.23</code>; or else are assumed to be columns 5,6,7. If present, a column named &#39;ngenes&#39; will be used to get the number of loci used to estimate the CFs for each 4-taxon set.</p><p>Output: <a href="#SNaQ.DataCF"><code>DataCF</code></a> object</p><p>Optional arguments:</p><ul><li><code>summaryfile</code>: if specified, a summary file will be created with that name.</li><li><code>delim</code> (for the first form only): to specify how columns are delimited, with single quotes: delim=&#39;;&#39;. Default is a <code>csv</code> file, i.e. <code>delim=&#39;,&#39;</code>.</li><li><code>mergerows</code>: false by default. When true, will attempt to merge multiple rows corresponding to the same four-taxon set (by averaging their quartet CFs) even if none of the species is repeated within any row (that is, in any set of 4 taxa)</li></ul><p>The last version modifies the input data frame, if species are represented by multiple alleles for instance (see <a href="#SNaQ.readtableCF!-Tuple{DataCF, DataFrames.DataFrame, Vector{Int64}}"><code>readtableCF!</code></a>(data frame, columns)).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readquartetdata.jl#L68-L96">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.readtrees2CF-Tuple{AbstractString}" href="#SNaQ.readtrees2CF-Tuple{AbstractString}"><code>SNaQ.readtrees2CF</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readtrees2CF(treefile)
readtrees2CF(vector of trees)</code></pre><p>Read trees in parenthetical format from a file, or take a vector of trees already read, and calculate the proportion of these trees having a given quartet (concordance factor: CF), for all quartets or for a sample of quartets. Optional arguments include:</p><ul><li><code>quartetfile</code>: name of text file with list of 4-taxon subsets to be analyzed. If none is specified, the function will list all possible 4-taxon subsets.</li><li><code>whichQ=&quot;rand&quot;</code>: to choose a random sample of 4-taxon subsets</li><li><code>numQ</code>: size of random sample (ignored if whichQ is not set to &quot;rand&quot;)</li><li><code>writeTab=false</code>: does not write the observedCF to a table (default true)</li><li><code>CFfile</code>: name of file to save the observedCF (default tableCF.txt)</li><li><code>writeQ=true</code>: save intermediate files with the list of all 4-taxon subsets and chosen random sample (default false).</li><li><code>writeSummary</code>: write descriptive stats of input data (default: true)</li><li><code>nexus</code>: if true, it assumes the gene trees are written in nexus file (default: false)</li></ul><p>See also: <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.countquartetsintrees"><code>PhyloNetworks.countquartetsintrees</code></a>, which uses a much faster algorithm; <a href="#SNaQ.readtableCF-Tuple{AbstractString}"><code>readtableCF</code></a> to read a table of quartet CFs directly.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readquartetdata.jl#L676-L697">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.snaq!-Tuple{HybridNetwork, DataCF}" href="#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.snaq!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">snaq!(T::HybridNetwork, d::DataCF)</code></pre><p>Estimate the network (or tree) to fit observed quartet concordance factors (CFs) stored in a DataCF object, using maximum pseudo-likelihood. A level-1 network is assumed. The search starts from topology <code>T</code>, which can be a tree or a network with no more than <code>hmax</code> hybrid nodes. The function name ends with ! because it modifies the CF data <code>d</code> by updating its attributes <code>expCF</code>: CFs expected under the network model. It does <em>not</em> modify <code>T</code>. The quartet pseudo-deviance is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0.</p><p>Output:</p><ul><li>estimated network in file <code>.out</code> (also in <code>.log</code>): best network overall and list of networks from each individual run.</li><li>the best network and modifications of it, in file <code>.networks</code>. All networks in this file have the same undirected topology as the best network, but have different hybrid/gene flow directions. These other networks are reported with their pseudo-likelihood scores, because non-identifiability issues can cause them to have very similar scores, and because SNaQ was shown to estimate the undirected topology accurately but not the direction of hybridization in cases of near non-identifiability.</li><li>if any error occurred, file <code>.err</code> provides information (seed) to reproduce the error.</li></ul><p>There are many optional keyword arguments, including</p><ul><li><code>hmax</code> (default 1): maximum number of hybridizations allowed</li><li><code>propQuartets</code> (default 1): the proportion of observed quartet concordance factors in <code>d</code> to use when calculating network pseudo-likelihoods. Smaller values will lead to faster method runtime but may come at the expense of accuracy if lowered too far.</li><li><code>probQR</code> (default 0): the probability at any given step to use weighted random sampling of quartets when deciding where to make topological moves when proposing the next candidate network.</li><li><code>verbose</code> (default false): if true, print information about the numerical optimization</li><li><code>runs</code> (default 10): number of independent starting points for the search</li><li><code>outgroup</code> (default none): outgroup taxon to root the estimated topology at the very end</li><li><code>filename</code> (default &quot;snaq&quot;): root name for the output files (<code>.out</code>, <code>.err</code>). If empty (&quot;&quot;), files are <em>not</em> created, progress log goes to the screen only (standard out).</li><li><code>seed</code> (default 0 to get it from the clock): seed to replicate a given search</li><li><code>probST</code> (default 0.3): probability to start from <code>T</code> at each given run. With problability 1-probST, the search is started from an NNI modification of <code>T</code> along a tree edge with no hybrid neighbor, with a possible modification of one reticulation if <code>T</code> has one.</li><li><code>updateBL</code> (default true): If true and if <code>T</code> is a tree, the branch lengths in <code>T</code> are first optimized roughly with <a href="../internals/#SNaQ.updateBL!-Tuple{HybridNetwork, DataCF}"><code>updateBL!</code></a> by using the average CF of all quartets defining each branch and back-calculating the coalescent units.</li></ul><p>The following optional keyword arguments control when to stop the optimization of branch lengths and γ&#39;s on each individual candidate network. Defaults are in parentheses:</p><ul><li><code>ftolRel</code> (1e-6) and <code>ftolAbs</code> (1e-6): relative and absolute differences of the network score between the current and proposed parameters,</li><li><code>xtolRel</code> (1e-2) and <code>xtolAbs</code> (1e-3): relative and absolute differences between the current and proposed parameters.</li></ul><p>Greater values will result in a less thorough but faster search. These parameters are used when evaluating candidate networks only. The following optional keyword arguments control when to stop proposing new network topologies:</p><ul><li><code>Nfail</code> (75): maximum number of times that new topologies are proposed and rejected (in a row).</li><li><code>liktolAbs</code> (1e-6): the proposed network is accepted if its score is better than the current score by at least liktolAbs.</li></ul><p>Lower values of <code>Nfail</code> and greater values of <code>liktolAbs</code> and <code>ftolAbs</code> would result in a less thorough but faster search.</p><p>At the end, branch lengths and γ&#39;s are optimized on the last &quot;best&quot; network with different and very thorough tolerance parameters: 1e-12 for <code>ftolRel</code>, 1e-10 for <code>ftolAbs</code>, <code>xtolRel</code>, <code>xtolAbs</code>.</p><p>The following optional keyword arguments are used to identify and exclude uninformative quartets. Uninformative quartets are those with concordance factors sufficiently close to the expected concordance factors from the star tree (one-third for all topologies).  Default parameters are are in parentheses:</p><ul><li><code>qinfTest</code> (false): if true, then look for uninformative quartets to discard.</li><li><code>qtolAbs</code> (1e-4): The tolerance for identifying uninformative concordance factors. Uninformative concordance factors are (1/3)±<code>qtolAbs</code></li></ul><p>See also: <a href="#SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>topologymaxQpseudolik!</code></a> to optimize parameters on a fixed topology, and <a href="#SNaQ.topologyQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>topologyQpseudolik!</code></a> to get the deviance (pseudo log-likelihood up to a constant) of a fixed topology with fixed parameters.</p><p>References:</p><p>Claudia Solís-Lemus and Cécile Ané (2016). Inferring phylogenetic networks with maximum pseudolikelihood under incomplete lineage sorting. <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005896">PLoS Genetics 12(3):e1005896</a></p><p>Kolbow, N, Kong, K, Chafin, T, Justison, J, Ane, C, Solis-Lemus, C (2025). SNaQ.jl: Improved scalability for phylogenetic network inference.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/snaq_optimization.jl#L1945-L2036">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.summarizedataCF-Tuple{DataCF}" href="#SNaQ.summarizedataCF-Tuple{DataCF}"><code>SNaQ.summarizedataCF</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">summarizedataCF(d::DataCF)</code></pre><p>function to summarize the information contained in a DataCF object. It has the following optional arguments:</p><ul><li><code>filename</code>: if provided, the summary will be saved in the filename, not to screen</li><li><code>pc</code> (number between (0,1)): threshold of percentage of missing genes to identify 4-taxon subsets with fewer genes than the threshold</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/readquartetdata.jl#L808-L814">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.topologyQpseudolik!-Tuple{HybridNetwork, DataCF}" href="#SNaQ.topologyQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.topologyQpseudolik!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">topologyQpseudolik!(net::HybridNetwork, d::DataCF)</code></pre><p>Calculate the quartet pseudo-deviance of a given network/tree for DataCF <code>d</code>. This is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0.</p><p>Note that the network&#39;s parameters (edge lengths and γs) are <em>not</em> optimized. So be careful if <code>net</code> does not have all internal branch lengths specified, because then the pseudolikelihood will be meaningless. See <a href="#SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>topologymaxQpseudolik!</code></a> if you want branch lengths and numerical parameters optimized on the given network.</p><p>The <code>fscore</code> field of the network is updated to contain the pseudo-deviance, and <code>d</code> is updated with the expected concordance factors under the input network.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/pseudolik.jl#L1326-L1341">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}" href="#SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.topologymaxQpseudolik!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">topologymaxQpseudolik!(
    net::HybridNetwork,
    d::DataCF;
    verbose = true,
    ftolRel = 1e-6,
    ftolAbs = 1e-6,
    xtolRel = 1e-2,
    xtolAbs = 1e-3
)</code></pre><p>Estimate the branch lengths and inheritance probabilities (γ&#39;s) for a given network topology. The network is <em>not</em> modified, only the object <code>d</code> is, with updated expected concordance factors.</p><p>Ouput: new network, with optimized parameters (branch lengths and gammas). The maximized quartet pseudo-deviance is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0. This is also an attribute of the network, which can be accessed with <code>loglik(net)</code>.</p><p>Optional arguments:</p><ul><li><code>verbose</code>: if <code>true</code>, information on the numerical optimization is printed to screen</li><li><code>ftolRel</code>, <code>ftolAbs</code>, <code>xtolRel</code>, <code>xtolAbs</code>: absolute and relative tolerance values for the pseudo-deviance function (<code>f</code>) and the parameters (<code>x</code>).</li></ul><p>The default tolerance values are quite lenient, for faster running time. They are more lenient than those used in <code>snaq!</code>, so we can expect that <code>snaq!</code> returns a better (lower) score for the same network topology. It is <em>highly</em> recommended to use more stringent value than the default, for example 1e-12 for <code>ftolRel</code>, and 1e-10 for <code>ftolAbs</code>, <code>xtolRel</code>, <code>xtolAbs</code> to match those in <a href="#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>snaq!</code></a>.</p><p>To further optimize branch lengths and γs, another strategy is the run the <code>topologymaxQpseudolik!</code> multiple times, because each time the edge parameters in the network are improved, and re-starting a search from a good place leads to finding even better edge parameter values. If <code>snaq!</code> finds a better score for the given network topology, then using this strategy (effectively used by <code>snaq!</code> when optimizing edge parameters at each trial) and using stringent tolerances should eliminate the difference.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/snaq_optimization.jl#L397-L437">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.DataCF" href="#SNaQ.DataCF"><code>SNaQ.DataCF</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">DataCF</code></pre><p>type that contains the following attributes:</p><ul><li><code>quartet</code> (vector of Quartets)</li><li><code>numQuartets</code></li><li><code>tree</code> (vector of trees: empty if a table of CF was input instead of list of trees)</li><li><code>numTrees</code> (-1 if a table CF was input instead of list of trees)</li><li><code>repSpecies</code> (taxon names that were repeated in table of CF or input gene trees: used inside snaq for multiple alleles case)</li></ul><p>The list of <code>Quartet</code> may be accessed with the attribute <code>.quartet</code>. If the input was a list of trees, the <code>HybridNetwork</code>&#39;s can be accessed with the attribute <code>.tree</code>. For example, if the <code>DataCF</code> object is named <code>d</code>, <code>d.quartet[1]</code> will show the first quartet and <code>d.tree[1]</code> will print the first input tree.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/types.jl#L248-L263">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SNaQ.Quartet" href="#SNaQ.Quartet"><code>SNaQ.Quartet</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Quartet</code></pre><p>type that saves the information on a given 4-taxon subset. It contains the following attributes:</p><ul><li><code>number</code>: integer</li><li><code>taxon</code>: vector of taxon names, like t1 t2 t3 t4</li><li><code>obsCF</code>: vector of observed CF, in order 12|34, 13|24, 14|23</li><li><code>logPseudoLik</code>: log pseudo-likelihood of the quartet. 0.0 by default</li><li><code>ngenes</code>: number of gene trees used to compute the observed CF; -1.0 if unknown</li><li><code>qnet</code>: <a href="../internals/#SNaQ.QuartetNetwork"><code>QuartetNetwork</code></a>, which saves the expCF after snaq estimation to emphasize that the expCF depend on a specific network, not the data</li><li><code>deltaCF</code>: The sum of absolute differences between observed and expected CFs</li><li><code>sampled</code>: A boolean denoting whether the quartet is used in computing the likelihood</li><li><code>uninformative</code>: A boolean denoting whether the quartet is not sampling due to being uninformative</li></ul><p>see also: <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/internals/#PhyloNetworks.QuartetT"><code>PhyloNetworks.QuartetT</code></a> for quartet with data of user-defined type <code>T</code>, using a mapping between quartet indices and quartet taxa.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/0078a0bef86acd11c1e43f6ff50605435f5c2dbc/src/types.jl#L198-L216">source</a></section></article><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#SNaQ.DataCF"><code>SNaQ.DataCF</code></a></li><li><a href="#SNaQ.Quartet"><code>SNaQ.Quartet</code></a></li><li><a href="#SNaQ.bootsnaq-Tuple{HybridNetwork, Union{DataFrames.DataFrame, Vector{Vector{HybridNetwork}}}}"><code>SNaQ.bootsnaq</code></a></li><li><a href="#SNaQ.fittedquartetCF"><code>SNaQ.fittedquartetCF</code></a></li><li><a href="#SNaQ.mapallelesCFtable-Tuple{AbstractString, AbstractString}"><code>SNaQ.mapallelesCFtable</code></a></li><li><a href="#SNaQ.readmultinewicklevel1-Tuple{AbstractString}"><code>SNaQ.readmultinewicklevel1</code></a></li><li><a href="#SNaQ.readnewicklevel1-Tuple{AbstractString}"><code>SNaQ.readnewicklevel1</code></a></li><li><a href="#SNaQ.readsnaqnetwork-Tuple{AbstractString}"><code>SNaQ.readsnaqnetwork</code></a></li><li><a href="#SNaQ.readtableCF-Tuple{AbstractString}"><code>SNaQ.readtableCF</code></a></li><li><a href="#SNaQ.readtableCF!-Tuple{DataCF, DataFrames.DataFrame, Vector{Int64}}"><code>SNaQ.readtableCF!</code></a></li><li><a href="#SNaQ.readtrees2CF-Tuple{AbstractString}"><code>SNaQ.readtrees2CF</code></a></li><li><a href="#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.snaq!</code></a></li><li><a href="#SNaQ.summarizedataCF-Tuple{DataCF}"><code>SNaQ.summarizedataCF</code></a></li><li><a href="#SNaQ.topologyQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.topologyQpseudolik!</code></a></li><li><a href="#SNaQ.topologymaxQpseudolik!-Tuple{HybridNetwork, DataCF}"><code>SNaQ.topologymaxQpseudolik!</code></a></li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../man/error_reporting/">« Error reporting</a><a class="docs-footer-nextpage" href="../internals/">Internals »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.15.0 on <span class="colophon-date" title="Friday 31 October 2025 17:46">Friday 31 October 2025</span>. Using Julia version 1.12.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
