<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Network estimation · SNaQ.jl</title><meta name="title" content="Network estimation · SNaQ.jl"/><meta property="og:title" content="Network estimation · SNaQ.jl"/><meta property="twitter:title" content="Network estimation · SNaQ.jl"/><meta name="description" content="Documentation for SNaQ.jl."/><meta property="og:description" content="Documentation for SNaQ.jl."/><meta property="twitter:description" content="Documentation for SNaQ.jl."/><meta property="og:url" content="https://JuliaPhylo.github.io/SNaQ.jl/man/snaq_est/"/><meta property="twitter:url" content="https://JuliaPhylo.github.io/SNaQ.jl/man/snaq_est/"/><link rel="canonical" href="https://JuliaPhylo.github.io/SNaQ.jl/man/snaq_est/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="SNaQ.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SNaQ.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li class="is-active"><a class="tocitem" href>Network estimation</a><ul class="internal"><li><a class="tocitem" href="#Inputs-for-SNaQ"><span>Inputs for SNaQ</span></a></li><li><a class="tocitem" href="#Estimating-a-network"><span>Estimating a network</span></a></li></ul></li><li><a class="tocitem" href="../fixednetworkoptim/">Candidate networks</a></li><li><a class="tocitem" href="../expectedCFs/">Extract expected CFs</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../parallelcomputation/">Parallel computation</a></li><li><a class="tocitem" href="../multiplealleles/">Multiple alleles</a></li><li><a class="tocitem" href="../error_reporting/">Error reporting</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Network estimation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Network estimation</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/SNaQ.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/SNaQ.jl/blob/dev/docs/src/man/snaq_est.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Network-estimation"><a class="docs-heading-anchor" href="#Network-estimation">Network estimation</a><a id="Network-estimation-1"></a><a class="docs-heading-anchor-permalink" href="#Network-estimation" title="Permalink"></a></h1><p>SNaQ (v1.0) implements the statistical inference method in <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005896">Solís-Lemus &amp; Ané 2016</a>. The procedure involves a numerical optimization of branch lengths and inheritance probabilities and a heuristic search in the space of phylogenetic networks.</p><p>We suggest that you create a special directory for running these examples, where input files can be downloaded and where output files will be created (with estimated networks for instance). Enter this directory and run Julia from there.</p><blockquote><p>This documentation pertains to SNaQ v1.0 as originally described in <a href="https://doi.org/10.1371/journal.pgen.1005896">Solís-Lemus &amp; Ané (2016)</a></p></blockquote><h2 id="Inputs-for-SNaQ"><a class="docs-heading-anchor" href="#Inputs-for-SNaQ">Inputs for SNaQ</a><a id="Inputs-for-SNaQ-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs-for-SNaQ" title="Permalink"></a></h2><p>SNaQ uses has two main inputs for estimating phylogenetic networks: concordance factors (CFs) and a starting tree (or network).</p><h3 id="Concordance-factors"><a class="docs-heading-anchor" href="#Concordance-factors">Concordance factors</a><a id="Concordance-factors-1"></a><a class="docs-heading-anchor-permalink" href="#Concordance-factors" title="Permalink"></a></h3><p>Concordance factors denote frequency of each quartet topology  present among the gene trees, which can be estimated using <a href="http://mrbayes.sourceforge.net">MrBayes</a> or <a href="http://sco.h-its.org/exelixis/software.html">RAxML</a> for example. </p><p><a href="https://juliaphylo.github.io/PhyloUtilities/">PhyloUtilities</a> has a step-by-step tutorial to go from multiple sequence alignments to a table of quartet gene frequencies (concordance factors: CFs), through BUCKy (to integrate out gene tree uncertainty) or through RAxML (see also the <a href="https://solislemuslab.github.io/snaq-tutorial/">snaq tutorial</a>). This pipeline uses sequence alignments in MrBayes to estimate gene trees, then BUCKy to produce the table of estimated CFs and their credibility intervals. Both steps in this pipeline are parallelized. </p><div class="admonition is-info" id="input-quartet-CFs:-BUCKy-versus-gene-trees-5ba409f4a1bdac0c"><header class="admonition-header">input quartet CFs: BUCKy versus gene trees<a class="admonition-anchor" href="#input-quartet-CFs:-BUCKy-versus-gene-trees-5ba409f4a1bdac0c" title="Permalink"></a></header><div class="admonition-body"><p>When quartet CFs are estimated using BUCKy on each subset of 4 taxa, then error in gene trees is accounted for: BUCKy aims to estimate true gene tree discordance, beyond discordance that is due to uncertainty in estimated gene trees. This is achieved in a Bayesian framework, using a full posterior sample of trees for each individual gene. So this method of obtaining quartet CFs</p><ul><li>is <em>not</em> what is sometimes referred to as a &quot;summary&quot; method</li><li>is expected to be robust to gene tree estimation error.</li></ul><p>When quartet CFs are calculated based on a single tree per gene, then gene tree error is <em>not</em> accounted for. Estimation error in gene trees can cause gene tree discordance. Downstream analysis with SNaQ would need to invoke deep coalescence or introgression to account for this apparent discordance. So this method</p><ul><li>may be called a &quot;summary&quot; method</li><li>may be sensitive to gene tree estimation error.</li></ul></div></div><h4 id="CFs-from-gene-trees"><a class="docs-heading-anchor" href="#CFs-from-gene-trees">CFs from gene trees</a><a id="CFs-from-gene-trees-1"></a><a class="docs-heading-anchor-permalink" href="#CFs-from-gene-trees" title="Permalink"></a></h4><p>Suppose you have a file with a list of gene trees in parenthetical format called <code>raxmltrees.tre</code>. You can access the example file of input trees <a href="https://github.com/juliaphylo/SNaQ/blob/main/examples/raxmltrees.tre">here</a> (or <a href="https://raw.githubusercontent.com/juliaphylo/SNaQ/main/examples/raxmltrees.tre">here</a> for easier download).</p><p>Do not copy-paste into a &quot;smart&quot; text-editor. Instead, save the file directly into your working directory using &quot;save link as&quot; or &quot;download linked file as&quot;. This file contains 30 gene trees, each in parenthetical format on 6 taxa like this (with rounded branch lengths):</p><p><code>(E:0.038,((A:0.014,B:0.010):0.010,(C:0.008,D:0.002):0.010):0.025,O:0.078);</code></p><p>If <code>raxmltrees.tre</code> is in your working directory, you can read in all gene trees and directly summarize them by a list of quartet CFs (proportion of input trees with a given quartet):</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using PhyloNetworks,SNaQ</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; raxmltrees = joinpath(dirname(pathof(SNaQ)), &quot;..&quot;,&quot;examples&quot;,&quot;raxmltrees.tre&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; raxmlCF = readtrees2CF(raxmltrees) # read in the file and produce a &quot;DataCF&quot; object</code><code class="nohighlight hljs ansi" style="display:block;">will use all quartets on 6 taxa
calculating obsCF from 30 gene trees and for 15 quartets
Reading in quartets...
0+---------------+100%  
  ***************  
table of obsCF printed to file tableCF.txt
descriptive stat of input data printed to file summaryTreesQuartets.txt
Object DataCF
number of quartets: 15
number of trees: 30</code></pre><p>In this table (<code>tableCF.txt</code>), each 4-taxon set is listed in one row. The 3 &quot;CF&quot; columns gives the proportion of genes that has each of the 3 possible trees on these 4 taxa.</p><p>When there are many more taxa, the number of quartets might be very large and we might want to use a subset to speed things up. Here, if we wanted to use a random sample of 10 quartets instead of all quartets, we could do:</p><p><code>raxmlCF = readtrees2CF(raxmltrees, whichQ=&quot;rand&quot;, numQ=10, CFfile=&quot;tableCF10.txt&quot;)</code></p><p>Be careful to use a <code>numQ</code> value smaller than the total number of possible 4-taxon subsets, which is <em>n choose 4</em> on <em>n</em> taxa (e.g. 15 on 6 taxa). To get a predictable random sample, you may set the seed with <code>using Random; Random.seed!(12321)</code> (for instance) prior to sampling the quartets as above. Additionally, providing a file name for the optional argument <code>CFile</code> saves the quartet CFs to file for later use.</p><h4 id="CFs-from-large-datasets"><a class="docs-heading-anchor" href="#CFs-from-large-datasets">CFs from large datasets</a><a id="CFs-from-large-datasets-1"></a><a class="docs-heading-anchor-permalink" href="#CFs-from-large-datasets" title="Permalink"></a></h4><p>When we want to get <em>all</em> quartet CFs,  the <a href="../../lib/public/#SNaQ.readtrees2CF-Tuple{AbstractString}"><code>readtrees2CF</code></a> is <em>much</em> slower than the PhyloNetworks function <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.countquartetsintrees"><code>PhyloNetworks.countquartetsintrees</code></a> to read in trees and calculate the quartet CFs observed in the trees:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; trees = readmultinewick(raxmltrees);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; q,t = countquartetsintrees(trees);</code><code class="nohighlight hljs ansi" style="display:block;">Reading in trees, looking at 15 quartets in each...
0+------------------------------+100%
  ******************************</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; nt = tablequartetCF(q,t); # named tuple</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; using DataFrames</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; df = DataFrame(nt, copycols=false); # convert to a data frame, without copying the column data</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; show(df, allcols=true) # data frames are displayed much more nicely than named tuples</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">15×9 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> qind   t1      t2      t3      t4      CF12_34   CF13_24    CF14_23    ngenes  </span>
     │<span class="sgr90"> Int64  String  String  String  String  Float64   Float64    Float64    Float64 </span>
─────┼────────────────────────────────────────────────────────────────────────────────
   1 │     1  A       B       C       D       1.0       0.0        0.0           30.0
   2 │     2  A       B       C       E       0.833333  0.0333333  0.133333      30.0
   3 │     3  A       B       D       E       0.833333  0.0333333  0.133333      30.0
   4 │     4  A       C       D       E       0.0       0.0        1.0           30.0
   5 │     5  B       C       D       E       0.0       0.0        1.0           30.0
   6 │     6  A       B       C       O       0.866667  0.0        0.133333      30.0
   7 │     7  A       B       D       O       0.866667  0.0        0.133333      30.0
   8 │     8  A       C       D       O       0.0       0.0        1.0           30.0
   9 │     9  B       C       D       O       0.0       0.0        1.0           30.0
  10 │    10  A       B       E       O       0.833333  0.0666667  0.1           30.0
  11 │    11  A       C       E       O       0.533333  0.333333   0.133333      30.0
  12 │    12  B       C       E       O       0.666667  0.266667   0.0666667     30.0
  13 │    13  A       D       E       O       0.533333  0.333333   0.133333      30.0
  14 │    14  B       D       E       O       0.666667  0.266667   0.0666667     30.0
  15 │    15  C       D       E       O       1.0       0.0        0.0           30.0</code></pre><h4 id="Reading-CFs-directly-from-file"><a class="docs-heading-anchor" href="#Reading-CFs-directly-from-file">Reading CFs directly from file</a><a id="Reading-CFs-directly-from-file-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-CFs-directly-from-file" title="Permalink"></a></h4><p>If we already have a table of quartet concordance factors (CFs) saved as a table in this format</p><table><tr><th style="text-align: left">Taxon1</th><th style="text-align: left">Taxon2</th><th style="text-align: left">Taxon3</th><th style="text-align: left">Taxon4</th><th style="text-align: left">CF12_34</th><th style="text-align: left">CF13_24</th><th style="text-align: left">CF14_23</th></tr><tr><td style="text-align: left">D</td><td style="text-align: left">A</td><td style="text-align: left">E</td><td style="text-align: left">O</td><td style="text-align: left">0.565</td><td style="text-align: left">0.0903</td><td style="text-align: left">0.3447</td></tr><tr><td style="text-align: left">...</td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left">...</td></tr></table><p>then we could read it in one step using the <a href="../../lib/public/#SNaQ.readtableCF-Tuple{AbstractString}"><code>readtableCF</code></a> function.</p><p>Concordance factors (CF), i.e. gene tree frequencies, for each 4-taxon subset can be obtained from <a href="http://www.stat.wisc.edu/~ane/bucky/">BUCKy</a> to account for gene tree uncertainty.</p><p>An example file comes with the package, available <a href="https://github.com/juliaphylo/SNaQ/blob/main/examples/buckyCF.csv">here</a> or <a href="https://raw.githubusercontent.com/juliaphylo/SNaQ/main/examples/buckyCF.csv">here</a>.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; buckyCFfile = joinpath(dirname(pathof(SNaQ)), &quot;..&quot;,&quot;examples&quot;,&quot;buckyCF.csv&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; buckyCF = readtableCF(buckyCFfile)</code><code class="nohighlight hljs ansi" style="display:block;">30.0 gene trees per 4-taxon set
Object DataCF
number of quartets: 15</code></pre><p>The same thing could be done in 2 steps: first to read the file and convert it to a <code>DataFrame</code> object from the <a href="https://dataframes.juliadata.org/stable/"><code>DataFrames</code></a> package and then to convert this <code>DataFrame</code> into a <a href="../../lib/public/#SNaQ.DataCF"><code>DataCF</code></a> object.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using CSV, DataFrames</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; dat = CSV.read(buckyCFfile, DataFrame);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; first(dat, 6) # to see the first 6 rows</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">6×14 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> taxon1   taxon2   taxon3   taxon4   CF12.34     CF12.34_lo  CF12.34_hi </span> ⋯
     │<span class="sgr90"> String1  String1  String1  String1  Float64     Float64     Float64    </span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ D        A        E        O        0.565033           0.5    0.633333  ⋯
   2 │ D        A        E        B        0.0005             0.0    0.0
   3 │ C        A        E        B        0.0005             0.0    0.0
   4 │ C        A        E        O        0.565033           0.5    0.633333
   5 │ D        A        O        B        3.33333e-5         0.0    0.0       ⋯
   6 │ C        D        E        O        0.999867           1.0    1.0
<span class="sgr36">                                                               7 columns omitted</span></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; buckyCF = readtableCF(dat)</code><code class="nohighlight hljs ansi" style="display:block;">30.0 gene trees per 4-taxon set
Object DataCF
number of quartets: 15</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; tablequartetCF(buckyCF)</code><code class="nohighlight hljs ansi" style="display:block;">(t1 = [&quot;D&quot;, &quot;D&quot;, &quot;C&quot;, &quot;C&quot;, &quot;D&quot;, &quot;C&quot;, &quot;C&quot;, &quot;C&quot;, &quot;C&quot;, &quot;C&quot;, &quot;D&quot;, &quot;C&quot;, &quot;C&quot;, &quot;A&quot;, &quot;C&quot;], t2 = [&quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;D&quot;, &quot;E&quot;, &quot;D&quot;, &quot;D&quot;, &quot;D&quot;, &quot;E&quot;, &quot;D&quot;, &quot;D&quot;, &quot;E&quot;, &quot;A&quot;], t3 = [&quot;E&quot;, &quot;E&quot;, &quot;E&quot;, &quot;E&quot;, &quot;O&quot;, &quot;E&quot;, &quot;O&quot;, &quot;A&quot;, &quot;O&quot;, &quot;E&quot;, &quot;O&quot;, &quot;A&quot;, &quot;A&quot;, &quot;O&quot;, &quot;O&quot;], t4 = [&quot;O&quot;, &quot;B&quot;, &quot;B&quot;, &quot;O&quot;, &quot;B&quot;, &quot;O&quot;, &quot;B&quot;, &quot;E&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;, &quot;O&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;], CF12_34 = [0.565033333333333, 0.0005, 0.0005, 0.565033333333333, 3.33333333333333e-5, 0.999866666666667, 0.0401666666666667, 1.0, 0.999866666666667, 1.0, 0.0401666666666667, 0.999866666666667, 1.0, 0.0731666666666667, 3.33333333333333e-5], CF13_24 = [0.0903, 0.8599, 0.8599, 0.0903, 0.8885, 6.66666666666667e-5, 0.263066666666667, 0.0, 6.66666666666667e-5, 0.0, 0.263066666666667, 6.66666666666667e-5, 0.0, 0.0424666666666667, 0.8885], CF14_23 = [0.344666666666667, 0.139566666666667, 0.139566666666667, 0.344666666666667, 0.111466666666667, 6.66666666666667e-5, 0.696766666666667, 0.0, 6.66666666666667e-5, 0.0, 0.696766666666667, 6.66666666666667e-5, 0.0, 0.884366666666667, 0.111466666666667], ngenes = Union{Missing, Float64}[30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0, 30.0])</code></pre><p>In the input file, columns need to be in the right order: with the first 4 columns giving the names of the taxa in each 4-taxon set. The CF values are assumed to be in columns named &quot;CF12_34&quot;, etc., or else in columns 5,6,7. If available, a column named &quot;ngenes&quot; will be taken to have the the number of genes for each 4-taxon subset.</p><h3 id="Starting-tree"><a class="docs-heading-anchor" href="#Starting-tree">Starting tree</a><a id="Starting-tree-1"></a><a class="docs-heading-anchor-permalink" href="#Starting-tree" title="Permalink"></a></h3><p>The other input for SNaQ is a starting tree (or network) to be used as a starting point in optimization.</p><p>If we have a tree for the data set at hand, it can be used as a starting point for the optimization. From our gene trees, we estimated a species tree with <a href="https://github.com/smirarab/ASTRAL/blob/master/astral-tutorial.md">ASTRAL</a>. This tree comes with the package in file <code>astral.tre</code> <a href="https://github.com/juliaphylo/SNaQ/blob/main/examples/astral.tre">here</a>. This file has 102 trees: 100 bootstrap species trees, followed by their greedy consensus, followed by the best tree on the original data. It&#39;s this last tree that we are most interested in. We can read it with</p><pre><code class="language-julia hljs">astralfile = joinpath(dirname(pathof(SNaQ)), &quot;..&quot;,&quot;examples&quot;,&quot;astral.tre&quot;);
astraltree = readmultinewick(astralfile)[102] # 102th tree: last tree here</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: E, O, C, D, ...
(E,(O,((C,D)100.0:3.029,(B,A)100.0:1.324)100.0:0.464));
</code></pre><p>Instead of a starting tree (<code>astraltree</code> in this case), we can start the optimization  in SNaQ with a network, but this network needs to be &quot;level-1&quot;. Note that all trees and all networks with 1 hybridization are of level 1. To make sure that a network with 2 or more hybridizations is of level 1, we can read it in with <code>readnewicklevel1</code> (which also unroots the tree, resolves polytomies, replaces missing branch lengths by 1 for starting values etc.):</p><pre><code class="language-julia hljs">T=readnewicklevel1(&quot;startNetwork.txt&quot;)</code></pre><p>Here <code>startNetwork.txt</code> is a hypothetical file: replace this by the name of a file that contains your network of interest.</p><h2 id="Estimating-a-network"><a class="docs-heading-anchor" href="#Estimating-a-network">Estimating a network</a><a id="Estimating-a-network-1"></a><a class="docs-heading-anchor-permalink" href="#Estimating-a-network" title="Permalink"></a></h2><p>After we have <a href="#Inputs-for-SNaQ">Inputs for SNaQ</a>, we can estimate the network using the input data <code>raxmlCF</code> and starting from tree (or network) <code>astraltree</code>. We first set <code>hmax=0</code> to impose the constraint of at most 0 hybrid node, that is, we estimate a tree.</p><pre><code class="language-julia hljs">net0 = snaq!(astraltree,raxmlCF, hmax=0, filename=&quot;net0&quot;, seed=1234)</code></pre><p>Part of the screen output shows this:</p><pre><code class="nohighlight hljs">MaxNet is (C,D,((B,A):1.395762055180493,(O,E):0.48453400554506426):10.0);
with -loglik 53.53150526187732</code></pre><p>This parenthetical (extended Newick) description is not very human-friendly, so we use <a href="https://github.com/juliaphylo/PhyloPlots.jl">PhyloPlots</a> to plot the tree (more about plotting networks can be found in the PhyloNetworks guide <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/net_plot/#Network-Visualization">Network Visualization</a>  and in <a href="https://github.com/juliaphylo/PhyloPlots.jl">PhyloPlots</a> documentation).</p><pre><code class="language-julia hljs">using PhyloPlots
plot(net0);</code></pre><p><img src="../../assets/figures/snaqplot_net0_1.svg" alt="net0_1"/></p><p>We can use this tree as a starting point to search for the best network allowing for at most <code>hmax=1</code> hybridizations (which is the default if we do not specify a <code>hmax</code> value).</p><pre><code class="language-julia hljs">net1 = snaq!(net0, raxmlCF, hmax=1, filename=&quot;net1&quot;, seed=2345)</code></pre><p>part of screen output:</p><pre><code class="nohighlight hljs">best network and networks with different hybrid/gene flow directions printed to .networks file
MaxNet is (C,D,((O,(E,#H7:::0.19558838614943078):0.31352437658618976):0.6640664399202987,(B,(A)#H7:::0.8044116138505693):10.0):10.0);
with -loglik 28.31506721890958</code></pre><p>We can visualize the estimated network and its inheritance values γ, which measure the proportion of genes inherited via each parent at a reticulation event (e.g. proportion of genes inherited via gene flow).</p><pre><code class="language-julia hljs">plot(net1, showgamma=true);</code></pre><p><img src="../../assets/figures/snaqplot_net1_1.svg" alt="net1_1"/></p><p>This network has A as a hybrid, 80.4% sister to B, and 19.6% sister to E (which is otherwise sister to O). C &amp; D are sister to each other.</p><p>Note that SNaQ infers an unrooted semi-directed network;  the lack of rooting is depicted visually with a trifurcationon on the leftmost side of the plot. The direction of hybrid edges can be inferred, but the direction of tree edges cannot be inferred. To obtain a representative visualization, it is best to root the network first, using one or more outgroup. PhyloNetworks has a guide on <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/dist_reroot/#Re-rooting-trees-and-networks">Re-rooting trees and networks</a> for this. If your outgroup conflicts with the direction of reticulations in the estimated network, see the section <a href="../fixednetworkoptim/#Candidate-networks-compatible-with-a-known-outgroup">Candidate networks compatible with a known outgroup</a>.</p><p>We can also check the output files created by <code>snaq!</code>:</p><pre><code class="language-julia hljs">less(&quot;net1.err&quot;) # would provide info about errors, if any
less(&quot;net1.out&quot;) # main output file with the estimated network from each run
less(&quot;net1.networks&quot;) # extra info</code></pre><p>when viewing these result files with <code>less</code> within Julia, use arrows to scroll down and type <code>q</code> to quit viewing the files.</p><ul><li><p>The <code>.networks</code> file contains a list of networks that are slight modifications of the best (estimated) network <code>net1</code>. The modifications change the direction of one reticulation at a time, by moving the placement of one hybrid node to another node inside the same cycle. For each modified network, the pseudolikelihood score was calculated (the <code>loglik</code> or <code>-Ploglik</code> values give a pseudo deviance actually).</p></li><li><p>The <code>.out</code> file contains the best network among all runs, and the best network per run, includes also the pseudolikelihood score and the computation time.</p></li><li><p>The <code>.log</code> file contains a description of each run, convergence criterion, and seed information.</p></li><li><p>The <code>.err</code> file has seed information on runs that failed, empty when nothing failed.</p></li></ul><p>The function name <a href="../../lib/public/#SNaQ.snaq!-Tuple{HybridNetwork, DataCF}"><code>snaq!</code></a> ends with ! because it modifies the argument <code>raxmlCF</code> by including the expected CF. Type <code>?</code> then <code>snaq!</code> to get help on that function.</p><p>The main output file, here <code>net1.out</code> (or <code>snaq.out</code> by default) has the estimated network in parenthetical format, but we can also print it directly to the screen:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; net1</code><code class="nohighlight hljs ansi" style="display:block;">HybridNetwork, Rooted Network
12 edges
12 nodes: 6 tips, 1 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(C,D,((O,(E,#H7:::0.196):0.314):0.664,((A)#H7:::0.804,B):10.0):10.0);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(net1)  # writes to screen, full precision for branch lengths and γ</code><code class="nohighlight hljs ansi" style="display:block;">&quot;(C,D,((O,(E,#H7:::0.19558838614943078):0.31352437658618976):0.6640664399202987,((A)#H7:::0.8044116138505693,B):10.0):10.0);&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(net1, round=true, digits=2)</code><code class="nohighlight hljs ansi" style="display:block;">&quot;(C,D,((O,(E,#H7:::0.2):0.31):0.66,((A)#H7:::0.8,B):10.0):10.0);&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(net1, di=true) # γ omitted: for dendroscope</code><code class="nohighlight hljs ansi" style="display:block;">&quot;(C,D,((O,(E,#H7):0.31352437658618976):0.6640664399202987,((A)#H7,B):10.0):10.0);&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(net1, &quot;bestnet_h1.tre&quot;) # writes to file: creates or overwrites file</code></pre><p>From a set of candidate networks, one might simply need to score of each network to pick the best. Here, the score is the negative log pseudo-deviance, and the lower the better. See the section to get the score of <a href="../fixednetworkoptim/#Candidate-networks">Candidate networks</a>.</p><h3 id="Choosing-the-number-of-hybridizations-(hmax)"><a class="docs-heading-anchor" href="#Choosing-the-number-of-hybridizations-(hmax)">Choosing the number of hybridizations (<code>hmax</code>)</a><a id="Choosing-the-number-of-hybridizations-(hmax)-1"></a><a class="docs-heading-anchor-permalink" href="#Choosing-the-number-of-hybridizations-(hmax)" title="Permalink"></a></h3><p>We change the <code>hmax</code> argument to change the search space to let the network have up to 2 or 3 hybrid nodes:</p><pre><code class="language-julia hljs">net2 = snaq!(net1,raxmlCF, hmax=2, filename=&quot;net2&quot;, seed=3456)
net3 = snaq!(net0,raxmlCF, hmax=3, filename=&quot;net3&quot;, seed=4567)</code></pre><p>and plot them (the optimized networks are identical and they both have a single reticulation):</p><pre><code class="language-julia hljs">using RCall                  # to be able to tweak our plot within R
R&quot;layout(matrix(1:2, 1, 2))&quot; # to get 2 plots into a single figure: 1 row, 2 columns
R&quot;par&quot;(mar=[0,0,1,0])        # for smaller margins
plot(net2, showgamma=true);
R&quot;mtext&quot;(&quot;hmax=2&quot;)           # add text annotation: title here
plot(net3, showgamma=true);
R&quot;mtext&quot;(&quot;hmax=3&quot;)</code></pre><p><img src="../../assets/figures/snaqplot_net23.svg" alt="net23"/></p><p>with this screen output for net2 (only 1 hybrid node found):</p><pre><code class="nohighlight hljs">MaxNet is (C,D,((B,(A)#H7:::0.804411606649347):10.0,(O,(#H7:::0.19558839335065303,E):0.3135243143217013):0.664066456871298):10.0);
with -loglik 28.31506721890957</code></pre><p>and this output for net3 (again, only 1 hybrid found):</p><pre><code class="nohighlight hljs">MaxNet is (D,C,((O,(E,#H7:::0.19558839257941849):0.3135243301652981):0.6640664138384673,(B,(A)#H7:::0.8044116074205815):10.0):10.0);
with -loglik 28.315067218909626</code></pre><p>Each network has a <code>loglik</code> attribute, which is its log pseudo deviance: a multiple of the negative log-likelihood up to a constant (the constant is such that the score is 0 if the network fits the data perfectly). The lower the better. We can plot these scores across hybrid values:</p><pre><code class="language-julia hljs">scores = [loglik(net0), loglik(net1), loglik(net2), loglik(net3)]
hmax = collect(0:3)
R&quot;plot&quot;(hmax, scores, type=&quot;b&quot;, ylab=&quot;network score&quot;, xlab=&quot;hmax&quot;, col=&quot;blue&quot;);</code></pre><p><img src="../../assets/figures/snaqplot_scores_heuristic.svg" alt="scores_heuristic"/></p><p>Here the slope heuristic suggests a single hybrid node: the score does not get much better beyond h=1.</p><p>We made the plot via R above. A more Julian way would use a Julia plotting package such as <a href="https://docs.juliaplots.org/latest/">Plots</a>, <a href="https://docs.makie.org/stable/tutorials/getting-started">Makie</a> or packages with similar principles to ggplot2 in R: <a href="https://aog.makie.org/v0.10.2/tutorials/intro-i">AlgebraOfGraphics</a> (based on Makie) or <a href="http://gadflyjl.org/stable/">Gadfly</a>. Also see this a cool (but now somewhat old) <a href="http://avt.im/blog/2018/03/23/R-packages-ggplot-in-julia">blog</a> about using ggplot within julia.</p><p>Note that since SNaQ assumes level-1 networks (i.e., no intersecting cycles), it might not be possible to add more hybridizations to networks with few taxa: </p><p><img src="../../assets/level1.png" alt="level1"/></p><h3 id="Suggestions-and-best-practices"><a class="docs-heading-anchor" href="#Suggestions-and-best-practices">Suggestions and best practices</a><a id="Suggestions-and-best-practices-1"></a><a class="docs-heading-anchor-permalink" href="#Suggestions-and-best-practices" title="Permalink"></a></h3><ul><li>For big datasets, do only one run (i.e., setting <code>runs=1</code>) first to get an idea of computing time:<pre><code class="language-julia hljs">net1 = snaq!(net0,raxmlCF, hmax=3, filename=&quot;net3&quot;, runs=1)</code></pre></li><li>Increase the number of hybridizations sequentially: <code>hmax=0,1,2,...</code>, and use the best network at <code>h-1</code> as starting point to estimate the best network at <code>h</code>.</li><li>Whenever possible, do many independent runs (default <code>runs=10</code>) to avoid getting stuck on local maxima. We do not need to do all runs sequentially. We can parallize by doing different runs on different cores (or computers). If we are on a machine or on a cluster that has many different cores, we can ask Julia to use these multiple cores, and <code>snaq!</code> will send different runs to different cores, like we did earlier.</li><li>For long jobs, run as a script in the terminal: <code>julia runSNaQ.jl</code>, arguments to the script are passed to Julia as a vector called <code>ARGS</code>. See the example script <code>runSNaQ.jl</code> in the <code>examples</code> folder of SNaQ, or the file located <a href="https://github.com/juliaphylo/SNaQ/blob/main/examples/runSNaQ.jl">here</a>. More on this topic in here: <a href="../parallelcomputation/#Parallel-computations">Parallel computations</a></li></ul><blockquote><p>This documentation pertains to SNaQ v1.0 as originally described in <a href="https://doi.org/10.1371/journal.pgen.1005896">Solís-Lemus &amp; Ané (2016)</a></p></blockquote></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../installation/">« Installation</a><a class="docs-footer-nextpage" href="../fixednetworkoptim/">Candidate networks »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Monday 11 August 2025 22:45">Monday 11 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
